test/state-component-mocha.js

```diff
@@ -0,0 +1,288 @@
+    it("shouldn't confuse a non-indexed port with undefined property result", function() {
+        var handler;
+        return Promise.resolve({
+            inPorts:{
+                a:{indexBy:'id'},
+                b:{indexBy:'id'},
+                c:{}
+            },
+            onchange: function(state, previously) {
+                handler(previously);
+                return _.keys(state);
+            }
+        }).then(stateComponent).then(createComponent).then(function(component){
+            // have the handler call a Promise resolve function to
+            // check that the data sent on the in port is passed to the handler
+            return new Promise(function(callback){
+                handler = callback;
+                sendData(component, 'a', {id:'undefined', d:"A"});
+            }).then(function(){
+                return new Promise(function(callback){
+                    handler = callback;
+                    sendData(component, 'c', {d:"C"});
+                });
+            });
+        }).should.become(undefined); // a should be absent because it's indexed
+    });
+    it("wait until all required ports have data", function() {
+        var handler;
+        return Promise.resolve({
+            inPorts:{
+                a: {required: true},
+                b: {required: true}
+            },
+            onchange: function(state) {
+                handler(state);
+            }
+        }).then(stateComponent).then(createComponent).then(function(component){
+            // have the handler call a Promise resolve function to
+            // check that the data sent on the in port is passed to the handler
+            return new Promise(function(callback){
+                handler = callback;
+                sendData(component, 'a', "A");
+                sendData(component, 'b', "B");
+            });
+        }).should.become({a: "A", b: "B"});
+    });
+    it("map indexed data together", function() {
+        var handler;
+        return Promise.resolve({
+            inPorts:{
+                a: {required: true, indexBy: 'aid'},
+                b: {required: true, indexBy: 'bid'}
+            },
+            onchange: function(state) {
+                handler(state);
+            }
+        }).then(stateComponent).then(createComponent).then(function(component){
+            // have the handler call a Promise resolve function to
+            // check that the data sent on the in port is passed to the handler
+            return new Promise(function(callback){
+                handler = callback;
+                sendData(component, 'a', {aid:"1", c: "A"});
+                sendData(component, 'a', {aid:"2", c: "A"});
+                sendData(component, 'a', {aid:"3", c: "A"});
+                sendData(component, 'b', {bid:"2", c: "B"});
+            });
+        }).should.become({a: {aid: "2", c: "A"}, b: {bid: "2", c: "B"}});
+    });
+    it("map indexed data by pointer", function() {
```

Comment:
How about something like "map indexed data by jsonpointer" ?   That lets someone new to jsonpointers know how to look it up.